crates/subspace-farmer/src/piece_cache.rs

```diff
@@ -230,35 +232,74 @@ where
         debug!("Collecting pieces that were in the cache before");
 
         // Build cache state of all backends
-        let mut caches = stored_pieces
-            .into_par_iter()
+        let maybe_caches_futures = stored_pieces
+            .into_iter()
             .zip(free_offsets)
             .zip(new_caches)
-            .map(|((mut stored_pieces, mut free_offsets), new_cache)| {
-                let contents = new_cache.contents();
-                stored_pieces.clear();
-                stored_pieces.reserve(contents.len());
-                free_offsets.clear();
-
-                for (offset, maybe_piece_index) in contents {
-                    match maybe_piece_index {
-                        Some(piece_index) => {
-                            stored_pieces
-                                .insert(RecordKey::from(piece_index.to_multihash()), offset);
-                        }
-                        None => {
-                            free_offsets.push(offset);
-                        }
-                    }
+            .enumerate()
+            .map(
+                |(index, ((mut stored_pieces, mut free_offsets), new_cache))| {
+                    run_future_in_dedicated_thread(
+                        move || async {
+                            let contents = new_cache.contents();
+                            stored_pieces.reserve(contents.len());
+
+                            for (offset, maybe_piece_index) in contents {
+                                match maybe_piece_index {
+                                    Some(piece_index) => {
+                                        stored_pieces.insert(
+                                            RecordKey::from(piece_index.to_multihash()),
+                                            offset,
+                                        );
+                                    }
+                                    None => {
+                                        free_offsets.push(offset);
+                                    }
+                                }
+
+                                // Allow for task to be aborted
+                                yield_now().await;
```

Comment:
We don't poll some shared cancellation channel here. Am I correct that by yielding the thread we "unblock" the process to shutdown? I would appreciate several more sentences of the description.

Reply:
This whole thing is a future. It just needs to yield to executor be abortable. This is exactly what we do here.